### PR TITLE
Runtime pin

### PR DESCRIPTION
**Pull request by HughNhan**
Description: 
This commit contains miscellaneous tweaks:
+ omit: a special postprocessing option to ignore first few seconds worth of results as warm-up period.
+ cpu-numa: option to pin client and server to same NUMA if possible, but owned CPU
+ runtime: compute runtime according to 'time' params taken 0-drop hunting iterations into account.

### CODE REVIEW

File: iperf-server-start [hunk 1]
```diff
@@ -176,31 +178,55 @@ ss -tlnp
...
+    # iperf PIN is to pin same NUMA and 1 CPU
+    IFS=',' read -r -a array <<< "$ifname_numa_node_cpus"
+    if [ "$1" == "client" ]; then
+        my_cpu=${array[${#array[@]}-1]}
+    else
+        my_cpu=${array[${#array[@]}-2]}
+    fi
+    cpu_pin_cmd="taskset --cpu-list ${my_cpu}"
+    echo $1: $cpu_pin_cmd
+}
```
Comment (k-rister) [inline iperf-server-start:214]: Since you have written this function twice now it seems like the kind of thing that should go in the `iperf-base` script.

**Review by k-rister — CHANGES_REQUESTED** [review]

File: iperf-client [hunk 2]
```diff
@@ -157,21 +164,37 @@ if [ -z "$remotehost" ]; then
...
-elif [ "${cpu_pin}" == "numa" ]; then
-    exit_error "cannot determine numa node for interface '${ifname}' and cpu-pin is numa"
+    if [ "$ifname_numa_node" == "-1" ]; then
+        # VM case, use numa 0
+        ifname_numa_node=0
+    fi
+else
+    # When topo=internode topo, the device is eth0. Default it to numa node0.
+    # In other word, if there is no numa associated with ifname, just use numa 0
+    ifname_numa_node=0
```
[the remark this answers is not in the document]
Comment (HughNhan) [reply iperf-client]: OK. Removed reference to topo specific.

File: iperf-client [hunk 3]
```diff
@@ -193,14 +216,35 @@ echo "time=$time"
...
+function assign-pin-cpu {
+    local my_cpu
+    # iperf PIN is to pin same NUMA and 1 CPU 
+    IFS=',' read -r -a array <<< "$ifname_numa_node_cpus"
+    if [ "$1" == "client" ]; then
+        my_cpu=${array[${#array[@]}-1]}
+    else
+        my_cpu=${array[${#array[@]}-2]}
+    fi
+	cpu_pin_cmd="taskset --cpu-list ${my_cpu}"
```
[the remark this answers is not in the document]
Comment (HughNhan) [reply iperf-client]: O> Fixed indentation, and added comment.

File: iperf-server-start [hunk 4]
```diff
@@ -150,7 +152,7 @@ if [ ! -z "$ifname" ]; then
         echo "IP was found: $ip"
     fi
 else
-    echo "Not going to look for an IP since --ifname was not used."
+    echo "Not going to look for an IP since --server-ifname was not used."
```
[the remark this answers is not in the document]
Comment (HughNhan) [reply iperf-server-start]: This was fixed.

File: iperf-server-start (hunk 1 above)
Comment (HughNhan) [reply iperf-server-start]: Good idea. Moved function to iperf-base.

File: iperf-client [hunk 5]
```diff
@@ -193,14 +216,35 @@ echo "time=$time"
...
+    echo $1: $cpu_pin_cmd
+}
+
 cpu_pin_cmd=""
 case "${cpu_pin}" in
     "numa")
 	cpu_pin_cmd="taskset --cpu-list ${ifname_numa_node_cpus}"
 	;;
+    "cpu-numa")
+    assign-pin-cpu client
```
[the remark this answers is not in the document]
Comment (HughNhan) [reply iperf-client]: Fixed.

**Review by k-rister — APPROVED** [review]

**Pull request merged**